### ISSUE TITLE
hi @karandesai

### ISSUE THREAD

**Issue by wjl0924** [karandesai-96/maxmin-cnn#3]
I had download your code ,but I can't find the realize maxmin-cnn model in detail. Can you shared the complete code ?

**Comment by karandesai-96** [comment]
@wjl0924 Hi, I did not get your complete question, you need to clone the repository, and run the jupyter notebook on your machine. The classical CNN layers - Convolution, Pooling, Fully Conencted / Dense layers have been provided by Keras libraries, while the **MaxMinConvolution** layer is inherited from Keras Convolution layer and the class code is written in **keras_maxmin_impl.py**. I am not sure what extra do you require here, could you clarify it a bit ?

If you are not acquainted with running Jupyter notebooks, you may get help here: http://jupyter-notebook-beginner-guide.readthedocs.io/en/latest/execute.html

**Comment by wjl0924** [comment]
oh，thank your clarify. I never used Jupyter notebooks, so I am in a mess.Let me have a look again.Thank you very much. By the way, Do you have to the authors of the paper to the source code?

**Comment by karandesai-96** [comment]
@wjl0924 Nice to see your doubt is clarified.

> By the way, Do you have to the authors of the paper to the source code?

I had a little chat with the authors, they haven't released their own source code yet. It has been written in CUDA with Torch 7 interface ( Lua language ).

**Comment by wjl0924** [comment]
oh,Here's another question,Does your code can use own data set? I want to use it to run myself a dataset.Thank you very much.

**Comment by wjl0924** [comment]
when I running
 `import matplotlib.pyplot as plt

# Changing default figure size for properly accomodating images
from pylab import rcParams
rcParams['figure.figsize'] = (16, 4)

# Show first 64 training images
for i in range(64):
    # Choosing a subplot to plot an image
    plt.subplot(4, 16, i + 1)

    # We don't need 0, 5, 10, 15... ticks on axes for images, removing them
    plt.xticks([])
    plt.yticks([])
    
    plt.imshow(X_train[i])
plt.show()`
 Why does the error message keep coming up?
Errors in ：
TypeError                                 Traceback (most recent call last)
<ipython-input-6-13b698406e44> in <module>()
     14     plt.yticks([])
     15 
---> 16     plt.imshow(X_train[i])
     17 plt.show()

/home/wjl/anaconda2/lib/python2.7/site-packages/matplotlib/pyplot.pyc in imshow(X, cmap, norm, aspect, interpolation, alpha, vmin, vmax, origin, extent, shape, filternorm, filterrad, imlim, resample, url, hold, data, **kwargs)
   3020                         filternorm=filternorm, filterrad=filterrad,
   3021                         imlim=imlim, resample=resample, url=url, data=data,
-> 3022                         **kwargs)
   3023     finally:
   3024         ax.hold(washold)

/home/wjl/anaconda2/lib/python2.7/site-packages/matplotlib/__init__.pyc in inner(ax, *args, **kwargs)
   1810                     warnings.warn(msg % (label_namer, func.__name__),
   1811                                   RuntimeWarning, stacklevel=2)
-> 1812             return func(ax, *args, **kwargs)
   1813         pre_doc = inner.__doc__
   1814         if pre_doc is None:

/home/wjl/anaconda2/lib/python2.7/site-packages/matplotlib/axes/_axes.pyc in imshow(self, X, cmap, norm, aspect, interpolation, alpha, vmin, vmax, origin, extent, shape, filternorm, filterrad, imlim, resample, url, **kwargs)
   4945                               resample=resample, **kwargs)
   4946 
-> 4947         im.set_data(X)
   4948         im.set_alpha(alpha)
   4949         if im.get_clip_path() is None:

/home/wjl/anaconda2/lib/python2.7/site-packages/matplotlib/image.pyc in set_data(self, A)
    451         if (self._A.ndim not in (2, 3) or
    452                 (self._A.ndim == 3 and self._A.shape[-1] not in (3, 4))):
--> 453             raise TypeError("Invalid dimensions for image data")
    454 
    455         self._imcache = None

TypeError: Invalid dimensions for image data

**Comment by karandesai-96** [comment]
> oh,Here's another question,Does your code can use own data set? I want to use it to run myself a dataset.Thank you very much.

Yes, you can but not the complete notebook, you can extract out the code where I declare the model, compile it and `fit` it. Please manage the variable names accordingly.

> Why does the error message keep coming up?

You need to pass a 2D or 3D numpy array to the method.  
Refer: http://matplotlib.org/api/pyplot_api.html#matplotlib.pyplot.imshow

If there's a problem in running my own code, you need to execute each block sequentially from the top one by one.

**Comment by wjl0924** [comment]
thank you ,I have others problem,when I running 
`from keras.models import Sequential
from keras.layers import Convolution2D,MaxPooling2D,Flatten,Dense
classical_cnn =Sequential([
     Convolution2D(32, 5, 5,activation='relu',border_mode='full',input_shape(3,32,32)),
     MaxPooling2D(pool_size=(3,3),strides(2,2)),
     
     Convolution2D(32, 5, 5,activation='relu',border_mode='full'),
     MaxPooling2D(pool_size=(3,3),strides(2,2)),
     
     Convolution2D(64, 5, 5,activation='relu',border_mode='full'),
     MaxPooling2D(pool_size=(3,3),strides(2,2)),
     
     Flatten(),
     Dense(64,activation='relu'),
     Dense(NB_CLASSES,activation='softmax'),
                           ])`
Errors in 
 File "<ipython-input-11-b4ae2254f4e1>", line 4
    Convolution2D(32, 5, 5,activation='relu',border_mode='full',input_shape(3,32,32)),
SyntaxError: non-keyword arg after keyword arg

**Comment by karandesai-96** [comment]
Hi @wjl0924 , it should be `input_shape=(3,32,32)`. It happened because you missed an equal to (`=`) sign. :smile: Let me know if there's any more problem in execution.

**Comment by wjl0924** [comment]
oh,thank you very much ,Your code has run cleanly.Currently, I am thinking ,how can  use own data set? 
Or there will be a lot of problems, Need to ask you. I hope you don't mind me trouble.😄😄😄

**Comment by karandesai-96** [comment]
@wjl0924 Sure, I'll help you out with running my network with your dataset. I'll reply after this weekend. Thanks :smile:

**Comment by wjl0924** [comment]
thank you 😄.I  figured out a way to reconstruction,
![if0z9q_v y dqnkucb kgv](https://cloud.githubusercontent.com/assets/21993898/21747766/afb24b50-d5ad-11e6-9339-d979b21b3445.png),
you can see The reconstructed pictures.That is  the final output into two parts.The top part of that function then updates the previous layer weights.The bottom part is still a softmax classifier.You can do to achieve the final output of this layer beasd on keras framework.

**Comment by wjl0924** [comment]
hello ,Have you got the message? You can do to achieve this structure? I am looking forward to your reply.Thanks.😄

**Comment by karandesai-96** [comment]
Hi @wjl0924, sorry for the unavailability. The picture looks impressive . You can implement it in Keras very easily. I am not sure whether you are askingg some doubt here. :smile:

**Comment by wjl0924** [comment]
Thank you reply.I'm not familiar with this kears framework. I don't know how to reconstruct it.
Can you help me to realize it? I should be most grateful if you can help me to solve this difficult problem.Thank you very much.

**Comment by WJFXQC** [comment]
> oh，thank your clarify. I never used Jupyter notebooks, so I am in a mess.Let me have a look again.Thank you very much. By the way, Do you have to the authors of the paper to the source code?
@wjl0924 I also this code,but I can,t realise this model(max-min CNN),Can you shared the complete code ?
